### PR TITLE
[Schedulers Refactoring] Phase 1: timesteps and scaling

### PR DESCRIPTION
**Pull request by anton-l**
As the refactor in #616 started getting too big to review in one go, I decided to approach this in a couple of smaller PRs that touch as few files as possible.

**Note**: it's easier to start reviewing from `pipeline_stable_diffusion.py`, as it shows the reasoning behind the API changes better.

Basic ideas that this PR implements:
* `scheduler.timesteps` are now **always** just integer (descending?) indices in range `[0, num_inference_steps)` in inference mode, e.g. `[49, 48,...,1, 0]`. This makes sure that we always iterate over the same range for every scheduler.
* `scheduler.schedule` replaces the original `timesteps` and contains either discrete noise conditions (`t` or `sigma`) for the model (like in DDIM: `[999, 978, ... 20, 0]`) or the resampled ones (like in LMS: `[999.0, 977.7, ... 0.0]`)
* `scheduler.scale_initial_noise()` scales the initial `torch.randn`, as sometimes (e.g. in Karras, LMS or VE schedulers) the initial noise is not `N(0, 1)`, but rather in `N(0, max_sigma^2)`. This function has to be applied after sampling the noise in the pipeline.
* `scheduler.scale_model_input(sample, step)` **has to be** applied for each UNet input sample, as sometimes the inputs need to be scaled (e.g. for Karras or LMS)
* `scheduler.get_noise_condition(step)` gets the noise condition (`t` or `sigma`) for the UNet. Sometimes the `t` needs to be scaled (e.g. in Karras+Euler), so this is implemented as a future-proof function that can access the scheduler's parameters.
* `scheduler.step()` must **always**  accept a value from `scheduler.timesteps` (rather than `scheduler.schedule`) as input, so that we can use it as index for `schedule`, `sigmas` or whatever the scheduler needs for the step.

TODO: merge the [Pytorch schedulers](https://github.com/huggingface/diffusers/pull/534) and rebase these changes on top

Coming up in the next PRs (Phase 2+): 
* `t` and `sigma` interchangeability to use any continuous scheduler with a discrete model and vise-versa 
* better (readable/intuitive) support for Nth order solvers that require combining multiple forward passes of the UNet

### CODE REVIEW

File: src/diffusers/schedulers/scheduling_utils.py [hunk 1]
```diff
@@ -37,6 +38,28 @@ class SchedulerOutput(BaseOutput):
...
+        return sample
+
+    @abc.abstractmethod
+    def get_noise_condition(self, step: int):
+        """
+        Returns the input noise condition for the model (e.g. `timestep` or `sigma`).
+        """
+        raise NotImplementedError("Scheduler must implement the `get_noise_condition` function.")
+
+
```
Comment (anton-l) [inline src/diffusers/schedulers/scheduling_utils.py:62]: This class combines the new required functions and ideally should be merged with `SchedulerMixin` (left standalone for easier reviewing for now).

**Comment by anton-l** [comment]
Not touching the VE schedulers yet, they will follow when we have convertable t<->sigmas. 
But overall: ready for review!

File: src/diffusers/schedulers/scheduling_lms_discrete.py [hunk 2]
```diff
@@ -92,16 +92,38 @@ def __init__(
         self.alphas = 1.0 - self.betas
         self.alphas_cumprod = np.cumprod(self.alphas, axis=0)
 
-        self.sigmas = ((1 - self.alphas_cumprod) / self.alphas_cumprod) ** 0.5
+        sigmas = ((1 - self.alphas_cumprod) / self.alphas_cumprod) ** 0.5
+        sigmas = sigmas[::-1].copy()
+        self.sigmas = np.concatenate([sigmas, [0.0]]).astype(np.float32)
```
Comment (anton-l) [inline src/diffusers/schedulers/scheduling_lms_discrete.py:97]: Fixes https://github.com/huggingface/diffusers/issues/454

File: src/diffusers/dependency_versions_table.py [hunk 3]
```diff
@@ -17,7 +17,6 @@
     "jaxlib": "jaxlib>=0.1.65,<=0.3.6",
     "modelcards": "modelcards>=0.1.4",
     "numpy": "numpy",
-    "onnxruntime": "onnxruntime",
```
Comment (pcuenca) [inline src/diffusers/dependency_versions_table.py:20]: Is this part of the refactor? Sorry, just trying to understand it better :)

File: src/diffusers/schedulers/scheduling_utils.py (hunk 1 above)
Comment (pcuenca) [reply src/diffusers/schedulers/scheduling_utils.py]: Yes, I agree to have it merged; otherwise schedulers need to inherit from both.

File: src/diffusers/schedulers/scheduling_utils.py [hunk 4]
```diff
@@ -37,6 +38,27 @@ class SchedulerOutput(BaseOutput):
...
+        Scales the model input (`sample`) to the correct range for the scheduler.
+        """
+        return sample
+
+    @abc.abstractmethod
+    def get_noise_condition(self, step: int):
+        """
+        Returns the input noise condition for the model (e.g. `timestep` or `sigma`).
+        """
+        raise NotImplementedError("Scheduler must implement the `get_noise_condition` function.")
```
Comment (pcuenca) [inline src/diffusers/schedulers/scheduling_utils.py:59]: However, should we default it to invoking `step()` rather than raising an exception? Then only the schedulers that need it would implement it. If you are reading the code for current schedulers, there's one thing less to worry about.

**Review by pcuenca — COMMENTED** [review]
Initially I thought the changes would result in more verbose code in the pipelines, but then I realized they just replace special cases with function calls, and it's much clearer this way.

I just left a few comments about details I may not be understanding properly.

File: src/diffusers/dependency_versions_table.py (hunk 3 above)
Comment (anton-l) [reply src/diffusers/dependency_versions_table.py:20]: No, just byproduct of not running `make clean` in my previous PR :sweat_smile:

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py [hunk 5]
```diff
@@ -42,6 +42,14 @@ class StableDiffusionPipeline(DiffusionPipeline):
             Model that extracts features from generated images to be used as inputs for the `safety_checker`.
     """
 
+    vae: AutoencoderKL
+    text_encoder: CLIPTextModel
+    tokenizer: CLIPTokenizer
+    unet: UNet2DConditionModel
+    scheduler: Union[DDIMScheduler, PNDMScheduler, LMSDiscreteScheduler]
+    safety_checker: StableDiffusionSafetyChecker
+    feature_extractor: CLIPFeatureExtractor
```
[the remark this answers is not in the document]
Comment (anton-l) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:51]: Just a helper for IDEs, otherwise these class attributes don't get recognized for class completion, since we don't set them in the `__init__` explicitly. I should probably address this in a separate PR, will remove them here for now :)

File: src/diffusers/schedulers/scheduling_utils.py [hunk 6]
```diff
@@ -37,6 +38,27 @@ class SchedulerOutput(BaseOutput):
     prev_sample: torch.FloatTensor
 
 
+class BaseScheduler(abc.ABC):
```
[the remark this answers is not in the document]
Comment (anton-l) [reply src/diffusers/schedulers/scheduling_utils.py:40]: Can't remove `SchedulerMixin` completely yet, as it's used by the Flax schedulers, so I guess we'll address it after the Flax refactoring

**Comment by anton-l** [comment]
Resolved the merge conflicts, ready for review again :)

**Comment by kashif** [comment]
does that mean that the `self.timesteps` are again floats?

File: src/diffusers/pipelines/ddim/pipeline_ddim.py [hunk 7]
```diff
@@ -82,14 +82,15 @@ def __call__(
         # set step values
         self.scheduler.set_timesteps(num_inference_steps)
 
-        for t in self.progress_bar(self.scheduler.timesteps):
+        for step in self.progress_bar(self.scheduler.timesteps):
             # 1. predict noise model_output
+            t = self.scheduler.get_noise_condition(step)
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/ddim/pipeline_ddim.py:87]: it's confusing to me that `get_noise_condition( )` return a `time` integer. Also why is this need for DDIM?

File: src/diffusers/schedulers/scheduling_ddim.py [hunk 8]
```diff
@@ -155,7 +155,8 @@ def __init__(
 
         # setable values
         self.num_inference_steps = None
-        self.timesteps = np.arange(0, num_train_timesteps)[::-1]
+        self.schedule = np.arange(0, num_train_timesteps)
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_ddim.py:158]: What do we need the `schedule` for?

File: src/diffusers/schedulers/scheduling_ddim.py [hunk 9]
```diff
@@ -240,6 +249,8 @@ def step(
         # - pred_sample_direction -> "direction pointing to x_t"
         # - pred_prev_sample -> "x_t-1"
 
+        timestep = self.schedule[timestep]
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_ddim.py:252]: Here I'm a bit lost - the code now here is more complex than it was before

Comment (patrickvonplaten) [reply src/diffusers/schedulers/scheduling_ddim.py:252]: Why are we now passing a different timestep to the scheduler than before that we then revert here?

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py [hunk 10]
```diff
@@ -46,6 +46,14 @@ class StableDiffusionPipeline(DiffusionPipeline):
             Model that extracts features from generated images to be used as inputs for the `safety_checker`.
     """
 
+    vae: AutoencoderKL
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:49]: this is a bit unrelated to the PR - let's try to put this in a different PR next time ;-)

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py [hunk 11]
```diff
@@ -230,14 +238,11 @@ def __call__(
             if latents.shape != latents_shape:
                 raise ValueError(f"Unexpected latents shape, got {latents.shape}, expected {latents_shape}")
         latents = latents.to(self.device)
+        latents = self.scheduler.scale_initial_noise(latents)
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:241]: Nice this makes sense to me!

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py [hunk 12]
```diff
@@ -247,13 +252,11 @@ def __call__(
...
-        for i, t in enumerate(self.progress_bar(self.scheduler.timesteps)):
+        for step in self.progress_bar(self.scheduler.timesteps):
             # expand the latents if we are doing classifier free guidance
             latent_model_input = torch.cat([latents] * 2) if do_classifier_free_guidance else latents
-            if isinstance(self.scheduler, LMSDiscreteScheduler):
-                sigma = self.scheduler.sigmas[i]
-                # the model input needs to be scaled to match the continuous ODE formulation in K-LMS
-                latent_model_input = latent_model_input / ((sigma**2 + 1) ** 0.5)
+            latent_model_input = self.scheduler.scale_model_input(latent_model_input, step)
+            t = self.scheduler.get_noise_condition(step)
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:259]: Here I'm lost -> this function is not intuitive to me. `t` != `noise_condition` for me. Not happy if we need to force schedulers to have this function

File: src/diffusers/schedulers/scheduling_lms_discrete.py [hunk 13]
```diff
@@ -99,13 +99,36 @@ def __init__(
...
+        Scales the model input (`sample`) to the correct range for the scheduler.
+        """
+        sigma = self.sigmas[self.num_inference_steps - step - 1]
+        return sample / ((sigma**2 + 1) ** 0.5)
+
+    def get_noise_condition(self, step: int):
+        """
+        Returns the input noise condition for a model.
+        """
+        return self.schedule[step]
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_lms_discrete.py:130]: No one-liner functions please

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py (hunk 11 above)
Comment (patrickvonplaten) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:241]: I don't think this works correctly at the moment because the sigmas are changed when doing `self.scheduler.set_timesteps`

File: src/diffusers/schedulers/scheduling_utils.py [hunk 14]
```diff
@@ -36,6 +37,38 @@ class SchedulerOutput(BaseOutput):
...
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_utils.py:40]:
```suggestion
class BaseScheduler:
```

Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_utils.py:40]: Not super happy about this. If some schedulers don't need a `scale_initial_noise` function we shouldn't force them to call it to get back a no-op, same with `scale_model_input`

File: src/diffusers/schedulers/scheduling_utils.py [hunk 15]
```diff
@@ -37,6 +38,27 @@ class SchedulerOutput(BaseOutput):
...
+        return noise
+
+    def scale_model_input(self, sample: torch.FloatTensor, step: int):
+        """
+        Scales the model input (`sample`) to the correct range for the scheduler.
+        """
+        return sample
+
+    @abc.abstractmethod
+    def get_noise_condition(self, step: int):
```
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply src/diffusers/schedulers/scheduling_utils.py:56]: Not very intuitive for me tbh

**Review by patrickvonplaten — COMMENTED** [review]
To be honest, I'm not super convinced by this design:

- Don't think we should force schedulers to call functions if they don't need to. Instead we should raise an error if a function is not called for a scheduler. This is also better for backwards comp
- Currently, I don't think the `BaseScheduler` provides much value and I would not be in favor of adding it. The functions: `scale_initial_noise` and `scale_model_input` are not required by DDIM or PNDM but just K-LMS. Just because K-LMS needs them doesn't warrant to add them as no-ops to DDIM and PNDM IMO. I think it's not good design to say "The default we put in a class every scheduler inherits from is to not scale and then schedulers that need to scale have to overwrite the method". This means if I add a new continuous scheduler now and forget to add `scale_initial_noise` then I'll get silent errors. Instead I want to get a big error
- I think we're changing too much with too little gains here. We're still starting from a design we have chosen and should try to reduce the mental energy it takes to go from the current design to the new design. Here we introduce **a lot** of new variables, function names:
`self.schedule`, `scale_initial_noise`, `scale_model_input`, `get_noise_condition`, the meaning of `step` is redefined, sometimes we pass `step` to `timesteps` sometimes to `schedule` => This is too much. Let's try to limit the new things to be learnt.
Ideas:
- I think we should focus much more on **global** differences between schedulers and derive simple logic from there. Why don't we just give each scheduler a class variable `"discrete"` or `"continous"` (I think every scheduler has to be one of the two). We can enforce this by checking in `SchedulerMixin` that every scheduler needs to have exactly one of the two. Then depending on whether it's `"continous"` or `"discrete"` we scale inputs or not. We can throw nice error messages in the `step(...)` function of the scheduler and the model forward that complain if variables are passed in the incorrect space.
- I don't think we really need to change `DDIM` or `PNDM`'s functionality, we just need to adapt K-LMS
- I think it would make a lot of sense to also add a `"trained_on_continous"` and `"trained_on_discrete"` to the model's config so that we can also throw a nice error in the model's forward if the wrong `dtype` is passed

**Comment by keturn** [comment]
There were some things here that had me scratching my head a bit too, but 

1. I don't understand the differences between the continuous and discrete approaches well enough. Maybe I should try reading Kerras's _Elucidating_ again? Or is there an example of a scheduler with the same underlying math written in both styles that I could read to compare?
2. I was extending significant credit on the basis of this being a "part one of two" PR, expecting that fields or methods that seem trivial now will get extended in interesting ways in Part Two.

**Comment by keturn** [comment]
a possibly-related thread on sampler/scheduler interface design is https://github.com/crowsonkb/k-diffusion/pull/23

**Comment by patrickvonplaten** [comment]
Alternative to this design: https://github.com/huggingface/diffusers/pull/711

**Comment by anton-l** [comment]
Superseded by #719
